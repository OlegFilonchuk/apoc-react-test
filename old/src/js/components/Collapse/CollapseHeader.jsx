import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { getDataTestElementProps, dataTestElementPropTypes } from '../../utils/dataTestElementPropUtils';

export const iconPositions = {
  left: 'left',
  right: 'right',
  none: 'none'
};

const rightIconTemplate = options => (
  <div className="sc-panel-utils">
    <span onClick={options.onClick} className={options.iconClass} />
  </div>
);

const leftIconTemplate = options => <span onClick={options.onClick} className={options.iconClass} />;

const onCustomContentClick = e => e.stopPropagation();

const CollapseHeader = ({
  iconStyles,
  iconPosition,
  isPanelButton,
  isClickDisabled,
  isCollapsed,
  onClick,
  title,
  customContent,
  className,
  displayCounter,
  countSelectedItems,
  ...restProps
}) => {
  const iconCollapsedClass = iconStyles.collapsed;
  const iconExpandedClass = iconStyles.expanded;

  const finalClassName = classNames('sc-panel-title', className);
  const iconClass = classNames('sc-collapse-icon sc-fonticon', {
    [iconCollapsedClass]: isCollapsed,
    [iconExpandedClass]: !isCollapsed
  });

  const headClick = isPanelButton && !isClickDisabled ? onClick : null;
  const iconClick = !isPanelButton && !isClickDisabled ? onClick : null;
  const headRole = isPanelButton ? 'button' : null;

  const leftIcon =
    iconPosition === iconPositions.left
      ? leftIconTemplate({
          iconClass,
          onClick: iconClick
        })
      : null;
  const rightIcon =
    iconPosition === iconPositions.right
      ? rightIconTemplate({
          iconClass,
          onClick: iconClick
        })
      : null;

  const collapseCustomContent = customContent ? (
    <div onClick={onCustomContentClick} className="collapse__custom-content">
      {customContent}
    </div>
  ) : null;
  const counterSection =
    displayCounter && countSelectedItems ? <h3 className="collapse__counter">({countSelectedItems})</h3> : null;

  return (
    <div onClick={headClick} role={headRole} className="sc-panel-heading" {...getDataTestElementProps(restProps)}>
      <div className={finalClassName}>
        {leftIcon}
        <h3>{title}</h3>
        {counterSection}
      </div>
      {collapseCustomContent}
      {rightIcon}
    </div>
  );
};

CollapseHeader.propTypes = {
  /**
   * Collapse head title
   */
  title: PropTypes.node.isRequired,

  /**
   * Is collapse closed (true default)
   */
  isCollapsed: PropTypes.bool.isRequired,

  /**
   * Is whole panel clickable (true default)
   */
  isPanelButton: PropTypes.bool.isRequired,

  /**
   * Icon position default (right)
   */
  iconPosition: PropTypes.oneOf([iconPositions.left, iconPositions.right, iconPositions.none]).isRequired,

  /**
   * Icon styles for expanded and collapsed state. By default it's caret.
   */
  iconStyles: PropTypes.shape({
    expanded: PropTypes.string,
    collapsed: PropTypes.string
  }).isRequired,

  /**
   * Ignores click
   */
  isClickDisabled: PropTypes.bool.isRequired,

  /**
   * What should a
   */
  onClick: PropTypes.func.isRequired,

  /**
   * Children for additional collapse content if needed
   */
  customContent: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),

  /**
   * Any custom class names
   */
  className: PropTypes.string,

  /**
   * If true, show counter for selected items
   */
  displayCounter: PropTypes.bool,

  /**
   * Number of selected items
   */
  countSelectedItems: PropTypes.number,

  ...dataTestElementPropTypes
};

CollapseHeader.defaultProps = {
  customContent: undefined,
  className: '',
  displayCounter: false,
  countSelectedItems: null
};

export default CollapseHeader;
