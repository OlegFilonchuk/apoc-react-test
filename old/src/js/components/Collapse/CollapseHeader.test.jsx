import React from 'react';
import { mount } from 'enzyme';

import CollapseHeader, { iconPositions } from './CollapseHeader';

const initialProps = {
  title: 'Foo Bar Panel',
  isCollapsed: true,
  isPanelButton: false,
  iconPosition: iconPositions.left,
  iconStyles: {
    expanded: 'foo-expanded-class-name',
    collapsed: 'bar-collapsed-class-name'
  },
  isClickDisabled: false,
  onClick: jest.fn(),
  'data-test-element': 'collapseHeaderDataTestElement'
};

const mountComponent = (props = {}, mountFn = mount) => mountFn(<CollapseHeader {...initialProps} {...props} />);

describe(`<${CollapseHeader.name}>`, () => {
  it('should render collapsed component with Foo title', () => {
    const component = mountComponent({
      title: 'Foo Panel Title',
      isCollapsed: true
    });

    expect(component).toMatchSnapshot();
  });

  it('should render expanded component with Bar title', () => {
    const component = mountComponent({
      title: 'Bar Panel Title',
      isCollapsed: false
    });

    expect(component).toMatchSnapshot();
  });

  it('should render icon on left', () => {
    const component = mountComponent({
      iconPosition: iconPositions.left
    });

    expect(component).toMatchSnapshot();
  });

  it('should render icon on right', () => {
    const component = mountComponent({
      iconPosition: iconPositions.right
    });

    expect(component).toMatchSnapshot();
  });

  it('should trigger click callback only when clicking on the icon', () => {
    // given
    const clickFn = jest.fn();
    const component = mountComponent({
      iconPosition: iconPositions.right,
      onClick: clickFn
    });

    const $icon = component.find('.sc-collapse-icon');
    const $title = component.find('.sc-panel-title');

    // when
    $icon.simulate('click');
    $title.simulate('click');

    // then
    expect(clickFn).toHaveBeenCalledTimes(1);
  });

  it('should trigger click callback when clicking either on the icon or title', () => {
    // given
    const clickFn = jest.fn();
    const component = mountComponent({
      iconPosition: iconPositions.right,
      onClick: clickFn,
      isPanelButton: true
    });

    const $icon = component.find('.sc-collapse-icon');
    const $title = component.find('.sc-panel-title');

    // when
    $icon.simulate('click');
    $title.simulate('click');

    // then
    expect(clickFn).toHaveBeenCalledTimes(2);
  });

  it('should not trigger click callback when handler is disabled', () => {
    // given
    const clickFn = jest.fn();
    const component = mountComponent({
      iconPosition: iconPositions.right,
      onClick: clickFn,
      isPanelButton: true,
      isClickDisabled: true
    });

    const $icon = component.find('.sc-collapse-icon');
    const $title = component.find('.sc-panel-title');

    // when
    $icon.simulate('click');
    $title.simulate('click');

    // then
    expect(clickFn).not.toHaveBeenCalled();
  });
});
