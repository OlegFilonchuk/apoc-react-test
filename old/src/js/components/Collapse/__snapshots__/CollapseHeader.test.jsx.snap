// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollapseHeader> should render collapsed component with Foo title 1`] = `
<CollapseHeader
  className=""
  countSelectedItems={null}
  data-test-element="collapseHeaderDataTestElement"
  displayCounter={false}
  iconPosition="left"
  iconStyles={
    Object {
      "collapsed": "bar-collapsed-class-name",
      "expanded": "foo-expanded-class-name",
    }
  }
  isClickDisabled={false}
  isCollapsed={true}
  isPanelButton={false}
  onClick={[Function]}
  title="Foo Panel Title"
>
  <div
    className="sc-panel-heading"
    data-test-element="collapseHeaderDataTestElement"
    onClick={null}
    role={null}
  >
    <div
      className="sc-panel-title"
    >
      <span
        className="sc-collapse-icon sc-fonticon bar-collapsed-class-name"
        onClick={[Function]}
      />
      <h3>
        Foo Panel Title
      </h3>
    </div>
  </div>
</CollapseHeader>
`;

exports[`<CollapseHeader> should render expanded component with Bar title 1`] = `
<CollapseHeader
  className=""
  countSelectedItems={null}
  data-test-element="collapseHeaderDataTestElement"
  displayCounter={false}
  iconPosition="left"
  iconStyles={
    Object {
      "collapsed": "bar-collapsed-class-name",
      "expanded": "foo-expanded-class-name",
    }
  }
  isClickDisabled={false}
  isCollapsed={false}
  isPanelButton={false}
  onClick={[Function]}
  title="Bar Panel Title"
>
  <div
    className="sc-panel-heading"
    data-test-element="collapseHeaderDataTestElement"
    onClick={null}
    role={null}
  >
    <div
      className="sc-panel-title"
    >
      <span
        className="sc-collapse-icon sc-fonticon foo-expanded-class-name"
        onClick={[Function]}
      />
      <h3>
        Bar Panel Title
      </h3>
    </div>
  </div>
</CollapseHeader>
`;

exports[`<CollapseHeader> should render icon on left 1`] = `
<CollapseHeader
  className=""
  countSelectedItems={null}
  data-test-element="collapseHeaderDataTestElement"
  displayCounter={false}
  iconPosition="left"
  iconStyles={
    Object {
      "collapsed": "bar-collapsed-class-name",
      "expanded": "foo-expanded-class-name",
    }
  }
  isClickDisabled={false}
  isCollapsed={true}
  isPanelButton={false}
  onClick={[Function]}
  title="Foo Bar Panel"
>
  <div
    className="sc-panel-heading"
    data-test-element="collapseHeaderDataTestElement"
    onClick={null}
    role={null}
  >
    <div
      className="sc-panel-title"
    >
      <span
        className="sc-collapse-icon sc-fonticon bar-collapsed-class-name"
        onClick={[Function]}
      />
      <h3>
        Foo Bar Panel
      </h3>
    </div>
  </div>
</CollapseHeader>
`;

exports[`<CollapseHeader> should render icon on right 1`] = `
<CollapseHeader
  className=""
  countSelectedItems={null}
  data-test-element="collapseHeaderDataTestElement"
  displayCounter={false}
  iconPosition="right"
  iconStyles={
    Object {
      "collapsed": "bar-collapsed-class-name",
      "expanded": "foo-expanded-class-name",
    }
  }
  isClickDisabled={false}
  isCollapsed={true}
  isPanelButton={false}
  onClick={[Function]}
  title="Foo Bar Panel"
>
  <div
    className="sc-panel-heading"
    data-test-element="collapseHeaderDataTestElement"
    onClick={null}
    role={null}
  >
    <div
      className="sc-panel-title"
    >
      <h3>
        Foo Bar Panel
      </h3>
    </div>
    <div
      className="sc-panel-utils"
    >
      <span
        className="sc-collapse-icon sc-fonticon bar-collapsed-class-name"
        onClick={[Function]}
      />
    </div>
  </div>
</CollapseHeader>
`;
